contents/performance/Performance.md

```diff
@@ -0,0 +1,104 @@
+
+This is not a guide on how to optimize that 3%, but rather, to teach you standard optimization techniques that you should apply in all of your code, so that you do not create sub-optimal codes (premature pessimization). If you want to find out more about optimizing that 3%, you can find more at [Other Resources](#other-resources) below.
+
+The 2 techniques you should know are:
+
+* Using appropriate Data Structures and Algorithms
+* Using Cache Friendly codes
+
+## Using Appropriate Data Structures and Algorithms
+
+Using appropriate Data Structures and Algorithms can improve the execution speed of your program trememdously. However, it requires time to learn and understand the nuances of each Data Structure and Algorithm and when to use them.
+
+A good example of how Data Structures and Algorithms can improve execution time is with [binary search](https://www.tutorialspoint.com/data_structures_algorithms/binary_search_algorithm.htm). When finding a particular element in a list, you need to search through the entire list. If it has a million entries, every find will require you to look through **1,000,000** entries. However, if you have a [sorted](https://www.tutorialspoint.com/data_structures_algorithms/sorting_algorithms.htm) list, you can use binary search to reduce the number of entries you have to look through to around **20**!
+
+If you are interested in learning more about Data Structures and Algorithms, you can learn more from the following resources:
+
+* [Geeksforgeeks.org](http://www.geeksforgeeks.org/data-structures/)
+* [Introduction to Algorithms (Book)](https://www.amazon.com/Introduction-Algorithms-3rd-MIT-Press/dp/0262033844)
+* [MIT online course](http://courses.csail.mit.edu/6.851/spring12/lectures/)
+* [topcoder.com](https://www.topcoder.com/community/data-science/data-science-tutorials/)
+
+There are more of such resources online. Just Google and you shall find.
+
+## Using Cache Friendly code
+
+Before we can talk about this, we need to know what the computer memory is. Computer memory has different components, [registers](https://en.wikipedia.org/wiki/Processor_register), [L1/L2/L3 cache](https://www.cs.umd.edu/class/fall2001/cmsc411/proj01/cache/cache.html), [RAM](https://en.wikipedia.org/wiki/Random-access_memory), and [disk](https://en.wikipedia.org/wiki/Hard_disk_drive) in order of their speed.
+
+> You can take a look at the extent of the difference of their access speeds [here](https://gist.github.com/jboner/2841832) <br>
+> You can also check out this [infographic](http://imgur.com/8LIwV4C)
+
+To understand this topic, you only need to know how the cache works. It is okay if you do not understand exactly how the other components, registers, RAM and disk work.
+
+The faster a cache is, the more expensive it is. This means that the L1 cache (a few kilobytes big) is smaller than the L2 cache (a few megabytes big) and so on, until the RAM (a few gigabytes big).
```

Comment:
Suggest adding a line at the start of this paragraph to explain what a cache is?